exec/java-exec/src/main/java/org/apache/drill/exec/resourcemgr/rmblobmgr/RMConsistentBlobStoreManager.java

```diff
@@ -0,0 +1,354 @@
+        throw new ResourceUnavailableException(String.format("Drillbit with UUID %s which is assigned to query " +
+            "doesn't have enough memory available. [Details: %s, AvailableMemory: %s, RequiredMemory: %s]", bitUUID,
+          currentAvailableMemory, memoryToReserve, exceptionStringBuilder.toString()));
+      }
+      // Update local ClusterState
+      bitAvailableResources.setMemoryInBytes(currentAvailableMemory - memoryToReserve);
+      currentClusterState.put(bitUUID, bitAvailableResources);
+
+      // Update local ForemanResourceUsage for foremanNode with this query resource ask
+      final NodeResources currentState = usageMapAcrossDrillbits.get(bitUUID);
+      long availableMemory = currentState.getMemoryInBytes();
+      currentState.setMemoryInBytes(availableMemory - memoryToReserve);
+      usageMapAcrossDrillbits.put(bitUUID, currentState);
+    }
+
+    // update the local ClusterStateBlob with new information
+    currentClusterBlob.setClusterState(currentClusterState);
+
+    // update the local ForemanQueueUsageBlob with final ForemanResourceUsage
+    currentUsage.setRunningCount(currentRunningCount + 1);
+    currentUsage.setForemanUsage(usageMapAcrossDrillbits);
+    allForemanUsage.put(foremanNode, currentUsage);
+    resourceUsageBlob.setAllForemanInfo(allForemanUsage);
+
+    // Update local blob cache
+    rmStateBlobs.put(ClusterStateBlob.NAME, currentClusterBlob);
+    rmStateBlobs.put(queueBlobName, resourceUsageBlob);
+
+    // Persist the new blobs to Zookeeper
+    if (!writeAllRMBlobs(rmStateBlobs)) {
+      logger.error("Failed to update the cluster state blob and queue blob for queue {} in a transaction",
+        selectedQueue.getQueueName());
+      throw new RMBlobUpdateException(String.format("Failed to update the cluster state blob and queue blob in a " +
+        "transaction. [Details: %s]", exceptionStringBuilder.toString()));
+    }
+    logger.debug("Successfully updated the blobs in a transaction. [Details: %s]", exceptionStringBuilder.toString());
+
+    // Reset the exceptionStringBuilder for next event
+    exceptionStringBuilder.delete(0, exceptionStringBuilder.length());
+  }
+
+  private void acquireLockAndUpdate(Map<String, NodeResources> queryResourceAssignment, QueryQueueConfig selectedQueue,
+                                   String leaderId, String queryId, String foremanNode) throws Exception {
+    try {
+      globalBlobMutex.acquire();
+    } catch (Exception ex) {
+      logger.error("Failed on acquiring the global mutex while updating the RM blobs during update of resources");
+      throw ex;
+    }
+
+    try {
+      updateBlobs(queryResourceAssignment, selectedQueue, leaderId, queryId, foremanNode);
+    } catch (Exception ex) {
+      logger.error("Failed to update the blobs", ex);
+      throw ex;
+    } finally {
+      // Check if the caller has acquired the mutex
+      if (globalBlobMutex.isAcquiredInThisProcess()) {
```

Comment:
why is this check required. Shouldn't be the case that we should acquire the lock if we are here?